bin/blocklock-agent/src/lib.rs

```diff
@@ -36,34 +37,52 @@ pub async fn run_agent<F, P>(
     agent: &mut BlocklockAgent<F, P>,
     ticker: NotifyTicker,
     decryption_sender_contract: DecryptionSender::DecryptionSenderInstance<P>,
+    block_poll_interval: std::time::Duration,
+    ticker_interval: std::time::Duration,
 ) -> anyhow::Result<()>
 where
     F: RequestChannel<Request = DecryptionRequest>,
     P: Provider + Clone + 'static,
 {
-    let mut events_stream = create_events_stream(decryption_sender_contract.clone()).await?;
-    loop {
-        match events_stream.next().await {
-            Some(ChainEvent::NewBlock(new_block)) => {
-                tracing::debug!(block_number = new_block, "ChainEvent::NewBlock");
-
-                // Update the blocklock state
-                agent.handle_new_block(new_block.into()).await;
-
-                // Tick the decryption fulfiller to try to satisfy requests every new block
+    let ticker_fut = {
+        let mut interval = tokio::time::interval(ticker_interval);
+        async move {
+            loop {
+                interval.tick().await;
                 ticker.0.notify_one();
             }
-            Some(ChainEvent::DecryptionRequested(request)) => {
-                tracing::info!(request_id = %request.requestId, "ChainEvent::DecryptionRequested");
-                agent.handle_decryption_requested(request).await;
-            }
-            None => Err(anyhow!("events stream ended prematurely"))?,
-        };
+        }
+    };
+
+    let mut events_stream =
+        create_events_stream(decryption_sender_contract.clone(), block_poll_interval).await?;
+    let events_loop = async move {
+        loop {
+            match events_stream.next().await {
+                Some(ChainEvent::NewBlock(new_block)) => {
+                    tracing::debug!(block_number = new_block, "ChainEvent::NewBlock");
+
+                    // Update the blocklock state
+                    agent.handle_new_block(new_block.into()).await;
+                }
+                Some(ChainEvent::DecryptionRequested(request)) => {
+                    tracing::info!(request_id = %request.requestId, "ChainEvent::DecryptionRequested");
+                    agent.handle_decryption_requested(request).await;
+                }
+                None => Err(anyhow!("events stream ended prematurely"))?,
+            };
+        }
+    };
+
+    tokio::select! {
+        out = events_loop => out,
+        _ = ticker_fut => unreachable!("ticker fut is always pending"),
```

Comment:
no such thing as a graceful shutdown?

Reply:
Fair, we could do a graceful termination by running the agent in a separate task and waiting